add support for skipping tasks - fixes #15

unkillbob opened the pull request:
![skipping](https://media.giphy.com/media/kEKcOWl8RMLde/giphy.gif)

Second attempt at skip task support (after https://github.com/SamVerschueren/listr/pull/17), this time with an explicit `skip()` function as per https://github.com/SamVerschueren/listr/pull/17#issuecomment-234362295.

> - if skip returns false or a Promise for false, continue execution.
> - if skip returns true or a Promise for true, skip with a generic message.
> - if skip returns a string or a Promise for a string, skip with a custom message.
> - if skip throws or returns a rejected promise, fail the entire build

An example of how this could be used in https://github.com/sindresorhus/np:

```js
{
	title: 'Publishing package',
	skip: () => {
		if (readPkgUp.sync().pkg.private) {
			return 'Private package: not publishing to npm.';
		}
	},
	task: () => exec('npm', ['publish'].concat(opts.tag ? ['--tag', opts.tag] : []))
}
```

<img width="319" alt="screen shot 2016-07-24 at 10 28 31 pm" src="https://cloud.githubusercontent.com/assets/1883524/17083263/02deacae-51ee-11e6-9da7-49227ee9adf5.png">

The tests could do with splitting up (probably to align nicely with the bulleted list of behaviour outlined above) and the readme needs updating but would be good to make sure everyone's happy with this solution before I get those done.

SamVerschueren commented on lib/renderer/string-renderer.js:
```diff
@@ -1,4 +1,7 @@
 'use strict';
+
```
remove empty line

SamVerschueren commented on lib/task.js:
```diff
@@ -46,11 +51,16 @@ class Task {
 			throw new TypeError(`Expected property \`task\` of type \`function\` but got \`${typeof task.task}\``);
 		}
 
+		if (typeof task.skip !== 'undefined' && typeof task.skip !== 'function') {
```
```js
if (task.skip && typeof task.skip !== 'function') {
```

SamVerschueren commented on lib/task.js:
```diff
@@ -80,6 +91,16 @@ class Task {
 				this.state = state.PENDING;
+				return this.skip();
+			})
+			.then(skip => {
```
`skip` -> `skipped`

SamVerschueren commented on lib/task.js:
```diff
@@ -80,6 +91,16 @@ class Task {
+			})
+			.then(skip => {
+				if (skip) {
+					const reason = typeof skip === 'string' ? skip : `Skipping ${this.title}.`;
```
I don't think we should print a reason if no reason was provided by the users themselves.

SamVerschueren commented on lib/task.js:
```diff
@@ -111,7 +132,9 @@ class Task {
 			})
 			.then(() => {
-				this.state = state.COMPLETED;
+				if (this.state !== state.SKIPPED) {
```
```js
if (this.state === state.PENDING) {
```

In case we add more states in the future.

SamVerschueren commented on test.js:
```diff
@@ -85,3 +85,45 @@ test('add tasks', t => {
 });
+
+test('skip tasks', t => {
+	t.plan(5);
```
remove `t.plan`

SamVerschueren commented on test.js:
```diff
@@ -85,3 +85,45 @@ test('add tasks', t => {
 	t.is(list._tasks.length, 4);
 });
+
+test('skip tasks', t => {
```
`test('skip tasks', async t => {`

SamVerschueren commented on test.js:
```diff
@@ -85,3 +85,45 @@ test('add tasks', t => {
+		}
+	]);
+
+	return list.run();
```
```js
await list.run();
```

SamVerschueren commented on test.js:
```diff
@@ -85,3 +85,45 @@ test('add tasks', t => {
+		{
+			title: 'Task 1',
+			task: () => new Listr([
+				{title: 'Task 1.1', task: () => t.pass() || '1.1'},
```
Why did you add the `|| '1.1'`?

unkillbob replied:
I had the `t.plan(5)` and the `t.pass()` to ensure that all the non-skipped tasks still ran. 

`t.pass()` doesn't return a value and currently the implementation of Listr requires the `task()` function return a result so that's why I returned *something* from each of the `task` functions.

But as mentioned in the PR description I'm keen to give the tests some love, this was just to prove the concept really.

SamVerschueren replied:
> I had the t.plan(5) and the t.pass() to ensure that all the non-skipped tasks still ran.

That's a valid point :). But can you change it to use `async`/`await` then? You can keep the `t.plan(5)` as that indeed makes sense.

> currently the implementation of Listr requires the task() function return a result

That indeed was the case but I fixed that yesterday as you can see [here](https://github.com/SamVerschueren/listr/blob/master/lib/task.js#L17-L19). It doesn't fail anymore if `result` is `undefined. So I believe it should work without the `|| '1.1';` stuff. If not, let me know because that should be fixed then.

> But as mentioned in the PR description I'm keen to give the tests some love, this was just to prove the concept really.

Very much appreciated :)!

SamVerschueren commented:
Super excited about this feature!

SamVerschueren commented:
Should we add `chalk.dim` to the `[Skipped]` part at the end of the line?

SamVerschueren commented on lib/task.js:
```diff
@@ -80,6 +91,17 @@ class Task {
+				if (skipped) {
+					this.state = state.SKIPPED;
+					if (typeof skipped === 'string') {
+						this._lastLine = stripAnsi(skipped.trim().split('\n').pop().trim());
```
Could you remove the `stripAnsi` call at this line and at line 125. I think it would make more sense to call it where the actual logging is done at line 83.

unkillbob commented:
@SamVerschueren I think this is ready for another look.

1. Tidied up and added more tests
2. Added skip docs to readme
3. Dimmed the `[skipped]` annotation in the title

SamVerschueren commented on readme.md:
```diff
@@ -135,6 +135,37 @@ const tasks = new Listr([
+	{
+		title: 'Can be skipped',
+		skip: () => {
+			if (someCondition) {
```
```js
if (Math.random() > 0.5) {
    return 'Reason for skipping';
}
```

SamVerschueren commented on readme.md:
```diff
@@ -135,6 +135,37 @@ const tasks = new Listr([
+			if (someCondition) {
+				return 'Reason for skipping';
+			}
+			return false;
```
Remove the `return false`

SamVerschueren commented:
We're almost there I guess. Can you put the task objects on different lines in the tests?

```js
{title: 'title', skip: () => false, task: () => { // task }}
```

=>

```js
{
    title: 'title',
    skip: () => false,
    task: () => {
        // task
    }
}
```

Much easier to reason about the logic.

@sindresorhus @jamestalmage Care to take a look at it?

sindresorhus commented:
LGTM

SamVerschueren commented on readme.md:
```diff
@@ -157,6 +187,12 @@ Type: `Function`
+
+Type: `Function`
+
+Skip function.
```
Should we show the different possibilities here? Otherwise it will be difficult for users to know what it does in the different scenario's.

> - if skip returns false or a Promise for false, continue execution.
> - if skip returns true or a Promise for true, skip with a generic message.
> - if skip returns a string or a Promise for a string, skip with a custom message.
> - if skip throws or returns a rejected promise, fail the entire build

After this, it's good to merge I guess 🎉  👍  ❤️

unkillbob replied:
I cover it in more detail on line 142, following the same pattern for the `task` function (which had more detail in the prose docs, and less detail in the API docs). Happy to change it, up to you 👍

SamVerschueren replied:
Oh yes. Maybe we should link to it then. 

> Read more about [skipping tasks]()

SamVerschueren commented:
Super great work @unkillbob ! Very much appreciate the effort 🍻
